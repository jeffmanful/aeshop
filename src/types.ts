export interface ProductEntity {
	id: number;
	title: string;
	description: string;
	img: string;
	price: string;
	seller: string;
	slug: string;
	size: string;
	brand: string;
	sold: boolean;
	liked: boolean;
}

export type Products = Array<ProductEntity>;

export interface ProductsState {
	products: Products;
}

export interface AddProductsAction {
	type: 'ADD_PRODUCTS';
	payload: ProductEntity[];
}

export interface ToggleProductAction {
	type: 'TOGGLE_PRODUCT_LIKED';
	id: number;
}

export type ProductActionTypes = AddProductsAction | ToggleProductAction;
