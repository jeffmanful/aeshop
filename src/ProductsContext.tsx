import React, { createContext, useEffect, useReducer, Dispatch } from 'react'
import { ProductsState, Products, ProductActionTypes } from './types'

const API_URL = ' https://5c78274f6810ec00148d0ff1.mockapi.io/api/v1/products'

export const initialState = {
  products: [],
}

export const ProductsContext = createContext<ProductsState>(initialState)
export const ReducerContext = createContext<Dispatch<ProductActionTypes>>(
  x => x
)

export const toggleLiked = (products: Products, id: number): Products =>
  products.map(product => (product.id !== id ? product : { ...product, liked: !product.liked }))

export function reducer(store: ProductsState, action: ProductActionTypes) {
  const { products } = store

  switch (action.type) {
    case 'TOGGLE_PRODUCT_LIKED':
      const { id } = action
      return { products: id ? toggleLiked(products, id) : [] }
    case 'ADD_PRODUCTS':
      const { payload } = action
      return { products: payload }
    default:
      return { products }
  }
}

export const ProductsProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  useEffect(() => {
    fetch(API_URL)
      .then(res => res.json())
      .then(result => {
        dispatch({
          type: 'ADD_PRODUCTS',
          payload: result,
        })
      })
      .catch(err => console.log('something went wrong', err))
  }, [])

  return (
    <ProductsContext.Provider value={state}>
      <ReducerContext.Provider value={dispatch}>
        {children}
      </ReducerContext.Provider>
    </ProductsContext.Provider>
  )
}
