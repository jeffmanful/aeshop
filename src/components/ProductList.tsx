import React, { useContext, useState, useEffect } from 'react'
import styled from 'styled-components'
import Box from './Box'
import Product from './Product'
import { ProductsContext } from '../ProductsContext'
import { ProductEntity, Products } from '../types'

const ProductListWrapper = styled.div`
  width: 95vw;
`

const ButtonFilter = styled.button`
  border-radius: 2px;
  color: black;
  border: 1px solid black;
  font-size: 1em;
  cursor: pointer;
  padding: 0.4rem 1rem;
  margin: 1rem;
  transition: 0.3s all;

  &:hover {
    color: white;
    background: black;
  }
`
ButtonFilter.displayName = 'ButtonFilter'

function filterBy(products: Array<ProductEntity>, type: string) {
  switch (type) {
    case 'HIDE_SOLD_ITEMS':
      return products.filter((product: ProductEntity) => !product.sold)
    case 'ALL':
    default:
      return products
  }
}

const ProductList = () => {
  const { products } = useContext(ProductsContext)
  const [filteredProducts, setFilteredProducts] = useState<Products>([])
  const [currentFilter, setCurrentFilter] = useState('ALL')

  useEffect(() => {
    setFilteredProducts(filterBy(products, currentFilter))
  }, [products, currentFilter])

  const productCount = filteredProducts?.length

  return (
    <ProductListWrapper data-testid="product-list">
      <div className="row">
        <p className="product-count">{productCount} styles</p>
        {currentFilter === 'ALL' ? (
          <ButtonFilter
            data-testid="hide-sold-items"
            onClick={() => setCurrentFilter('HIDE_SOLD_ITEMS')}
          >
            Hide sold items
          </ButtonFilter>
        ) : (
          <ButtonFilter
            data-testid="show-all-items"
            onClick={() => setCurrentFilter('ALL')}
          >
            Show all items
          </ButtonFilter>
        )}
      </div>

      <Box>
        {filteredProducts?.map((product: ProductEntity) => (
          <Product {...product} key={product.id} />
        ))}
      </Box>
    </ProductListWrapper>
  )
}
export default ProductList
