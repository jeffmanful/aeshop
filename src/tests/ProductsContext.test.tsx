import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { reducer, initialState } from '../ProductsContext'
import { ProductActionTypes, ProductsState } from '../types'
import '@testing-library/jest-dom/extend-expect'
import { state } from './ProductList.test'
import { ProductsProvider } from '../ProductsContext'
import ProductList from '../components/ProductList'
import { createMemoryHistory } from 'history'
import { Router } from 'react-router-dom'

let container: any = null

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

describe('ProductsProvider', () => {
  test('ProductsProvider fetch renders prodcutList', async () => {
    const history = createMemoryHistory()
    jest.spyOn<any, any>(global, 'fetch').mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(state.products),
      })
    )

    await act(async () => {
      render(
        <Router history={history}>
          <ProductsProvider>
            <ProductList />
          </ProductsProvider>
        </Router>,
        container
      )
    })

    expect(container.querySelector('.product-count').textContent).toBe(
      '3 styles'
    )
  })
})

describe('reducer', () => {
  test('returns passed products when there is no action type', () => {
    const action: ProductActionTypes = {
      type: 'TOGGLE_PRODUCT_LIKED',
      id: 1,
    }
    expect(reducer(initialState, action)).toEqual(initialState)
  })

  test('sets the product as liked', () => {
    const state: ProductsState = {
      products: [
        {
          id: 1,
          liked: false,
          price: '',
          description: '',
          title: '',
          img: '',
          sold: false,
          seller: '',
          slug: '',
        },
      ],
    }
    const action: ProductActionTypes = {
      type: 'TOGGLE_PRODUCT_LIKED',
      id: 1,
    }

    expect(state.products[0].liked).toEqual(false)

    const reducedState = reducer(state, action)
    const { products } = reducedState

    expect(products[0].liked).toEqual(true)
  })

  test('adds products to the store', () => {
    const action: ProductActionTypes = {
      type: 'ADD_PRODUCTS',
      payload: [],
    }

    const intialReducedState = reducer(initialState, action)
    expect(intialReducedState.products.length).toEqual(0)

    const actionWithPayload: ProductActionTypes = {
      type: 'ADD_PRODUCTS',
      payload: [
        {
          id: 1,
          liked: false,
          price: '',
          description: '',
          title: '',
          img: '',
          sold: false,
          seller: '',
          slug: '',
        },
        {
          id: 2,
          liked: false,
          price: '',
          description: '',
          title: '',
          img: '',
          sold: false,
          seller: '',
          slug: '',
        },
      ],
    }

    const reducedState = reducer(initialState, actionWithPayload)
    expect(reducedState.products.length).toEqual(2)
  })
})
