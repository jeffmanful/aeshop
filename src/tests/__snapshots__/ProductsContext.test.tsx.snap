// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxheI flMNKF"
    data-testid="product-list"
  >
    <div
      class="row"
    >
      <p
        class="product-count"
      >
        0 styles
      </p>
      <button
        class="sc-Axmtr gsHpch"
        data-testid="hide-sold-items"
      >
        Hide sold items
      </button>
    </div>
    <div
      class="sc-AxjAm FdrlR"
    />
  </div>
</DocumentFragment>
`;
